### PR TITLE
Introduce a function possibly_extend_to_N

### PR DESCRIPTION
**Pull request by class4kayaker**
Fix to issue #528 by adding a utility to create a vector of length N by either copying a given vec of length N or duplicating a single provided value N times. Commentary on implementation, error format/message would be welcome.

Currently only the double type is offered.

### CODE REVIEW

File: source/utilities.cc [hunk 1]
```diff
@@ -50,6 +50,45 @@ namespace aspect
...
+    std::vector<T>
+    possibly_extend_from_1_to_N (const std::vector<T> &values,
+                                 const unsigned int N,
+                                 const std::string &id_text)
+    {
+      std::vector<T> return_vector(N);
+      if (values.size() == 1)
+        {
+          for (unsigned int i=0; i<N; ++i)
+            return_vector[i] = values[0];
```
Comment (bangerth) [inline source/utilities.cc]: I think you can make this function simpler if you do something like
```
  if (values.size() == 1)
    return std::vector<T> (N, values[0]);
  else if (values.size() == N)
    return values;
  else
     // ...
    AssertThrow (....);

  // make compiler happy by returning something here, even though we should not
  // get here
  return std::vector<T> ();
```

**Comment by bangerth** [comment]
Nice simplification. I think the function (with the suggested modifications) is simple enough that you can even make it an inline function at the bottom of `utilities.h` (i.e., declare at the top, with comment, put the implementation at the bottom, with `inline`). This way you also don't have to worry about the instantiation in `utilities.cc`.

**Comment by class4kayaker** [comment]
I could not get the inline function working as you described on my machine. However, it may be that I did not know the correct format for the inline implementation in utilities.h.

File: source/utilities.cc [hunk 2]
```diff
@@ -50,6 +50,46 @@ namespace aspect
     using namespace dealii;
 
     /**
+     * Given an array @p values, consider three cases:
```
Comment (bangerth) [inline source/utilities.cc]: No need to repeat the documentation a second time here -- it is already fine in the .h file. If you have two copies of two things, they will quickly get out of synch.

File: source/utilities.cc [hunk 3]
```diff
@@ -50,6 +50,46 @@ namespace aspect
...
+     * contain different values for each of a set of objects (e.g., for
+     * each compositional field), or contain a single value that is then
+     * used for each object.
+     */
+    template <typename T>
+    std::vector<T>
+    possibly_extend_from_1_to_N (const std::vector<T> &values,
+                                 const unsigned int N,
+                                 const std::string &id_text)
+    {
```
Comment (bangerth) [inline source/utilities.cc]: just copy this whole implementation of the function into the .h file, somewhere towards the bottom.

**Comment by bangerth** [comment]
Yes, that's what I had in mind.

/run-tests

File: source/utilities.cc [hunk 4]
```diff
@@ -1247,6 +1247,10 @@ namespace aspect
     }
 
     // Explicit instantiations
+    template  std::vector<double> possibly_extend_from_1_to_N (const std::vector<double> &values,
+                                                               const unsigned int N,
+                                                               const std::string &id_text);
+
```
Comment (bangerth) [inline source/utilities.cc]: you don't need this any more -- just remove it

**Comment by bangerth** [comment]
In addition to the one last comment, can you rebase to master (will yield a conflict somewhere, so you'll need to resolve this) and squash the commits into one? Please say when done!

**Comment by bangerth** [comment]
OK to merge once the tester is through.

**Pull request merged**